optaplanner-core/src/main/java/org/optaplanner/core/impl/util/TreeMapValueList.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.util;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.TreeMap;
+
+public class TreeMapValueList<I, T> implements List<T> {
+    private final TreeMap<I, T> sourceMap;
+
+    public TreeMapValueList(TreeMap<I, T> sourceMap) {
+        this.sourceMap = sourceMap;
+    }
+
+    @Override
+    public int size() {
+        return sourceMap.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return sourceMap.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return sourceMap.values().contains(o);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return sourceMap.values().iterator();
+    }
+
+    @Override
+    public Object[] toArray() {
+        return sourceMap.values().toArray();
+    }
+
+    @Override
+    public <T1> T1[] toArray(T1[] a) {
+        return sourceMap.values().toArray(a);
+    }
+
+    @Override
+    public boolean add(T t) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        return sourceMap.values().containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends T> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends T> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c) {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
So all mutating methods are unsupported? In that case I doubt if the type is good for anything.

What's the benefit of this as opposed to a couple of static util methods?
Or a more tailored class that is not a `List`? It looks to me as if the `List` forces you to implement stuff you don't really need, obfuscating the purpose of the class.

Reply:
A `TreeMap` doesn't have a `get(index)` method. I expect most people wants a list of sequences (in the order the sequence occurs in). I also makes testing easier (as I can refer to the sequence by index). Most of this stuff would need to be implemented for any Collection type, leaving `Iterable` as the only interface we can implement so we get iterator functionality (but I expect most people want a Collection).